test/gtest/ucp/test_ucp_request.cc

```diff
@@ -448,47 +524,207 @@ UCS_TEST_P(test_proto_reset, am_eager_multi_bcopy, "ZCOPY_THRESH=inf",
+ * messages to be sent. */
+class test_proto_reset_atp : public test_proto_reset {
+public:
+    void init() override
+    {
+        test_proto_reset::init();
+        m_pending_reqs.clear();
+        m_atp_count = 0;
+    }
+
+private:
+    void hook_uct_cbs()
+    {
+        ucp_ep_h ep                = sender().ep();
+        ucp_lane_index_t num_lanes = ucp_ep_config(ep)->key.num_lanes;
+        uct_ep_h uct_ep;
+        uct_iface_ops_t *ops;
+
+        for (ucp_lane_index_t lane = 0; lane < num_lanes; ++lane) {
+            uct_ep                = ucp_ep_get_lane(ep, lane);
+            m_ops[uct_ep]         = uct_ep->iface->ops;
+            ops                   = &uct_ep->iface->ops;
+            ops->ep_pending_add   = add_pending;
+            ops->ep_pending_purge = purge_pending;
+            ops->ep_am_short      = uct_am_short;
+        }
+    }
+
+    void restore_uct_cbs()
+    {
+        ucp_ep_h ep                = sender().ep();
+        ucp_lane_index_t num_lanes = ucp_ep_config(ep)->key.num_lanes;
+
+        for (ucp_lane_index_t lane = 0; lane < num_lanes; ++lane) {
+            uct_ep_h uct_ep    = ucp_ep_get_lane(ep, lane);
+            uct_ep->iface->ops = m_ops[uct_ep];
+        }
+    }
+
+    static ucs_status_t uct_am_short(uct_ep_h ep, uint8_t id, uint64_t header,
+                                     const void *payload, unsigned length)
+    {
+        if ((id == UCP_AM_ID_RNDV_ATP) && (++m_atp_count == 2)) {
+            return UCS_ERR_NO_RESOURCE;
+        }
+
+        return m_ops[ep].ep_am_short(ep, id, header, payload, length);
+    }
+
+    static ucs_status_t
+    add_pending(uct_ep_h tl_ep, uct_pending_req_t *n, unsigned flag)
+    {
+        m_pending_reqs.push_back(n);
+        return UCS_OK;
+    }
+
+    static void
+    purge_pending(uct_ep_h ep, uct_pending_purge_callback_t cb, void *arg)
+    {
+        for (auto &req : m_pending_reqs) {
+            cb(req, arg);
+        }
+
+        m_pending_reqs.clear();
+    }
+
+protected:
+    void wait_and_restart(const std::vector<void*> &reqs) override
+    {
+        hook_uct_cbs();
+
+        /* Wait until first ATP was sent */
+        wait_any(reqs, [reqs](const void *ureq) {
+            if (ureq == NULL) {
+                return false;
+            }
+
+            ucp_request_t *req = (ucp_request_t*)ureq - 1;
+            return (req->flags & UCP_REQUEST_FLAG_PROTO_SEND) &&
+                   (req->send.proto_stage ==
+                    UCP_PROTO_RNDV_PUT_STAGE_FENCED_ATP);
+        });
+
+        restart(sender().ep());
+        restore_uct_cbs();
+    }
+
+    static unsigned m_atp_count;
+    static std::vector<uct_pending_req_t*> m_pending_reqs;
+    static std::map<uct_ep_h, uct_iface_ops> m_ops;
+};
+
+std::vector<uct_pending_req_t*> test_proto_reset_atp::m_pending_reqs;
+std::map<uct_ep_h, uct_iface_ops> test_proto_reset_atp::m_ops;
+unsigned test_proto_reset_atp::m_atp_count;
```

Comment:
why needed?

Reply:
compiler error otherwise:
`undefined reference to `test_proto_reset_atp::m_atp_count`

Reply:
ok